plugin/sign/signer.go

```diff
@@ -0,0 +1,218 @@
+			rrsig, err = pair.signRRs(z.Apex.NS, s.origin, s.ttl, s.inception, s.expiration)
+			if err != nil {
+				return nil, err
+			}
+			z.Insert(rrsig)
+		}
+		if apex {
+			rrsig, err = pair.signRRs([]dns.RR{nsec}, s.origin, s.ttl, s.inception, s.expiration)
+			if err != nil {
+				return nil, err
+			}
+			z.Insert(rrsig)
+		}
+	}
+
+	// We are walking the tree in the same direction, so names[] can be used here to indicated the next element.
+	i := 1
+	err = z.Walk(func(e *tree.Elem, zrrs map[uint16][]dns.RR) error {
+		if !apex && e.Name() == s.origin {
+			nsec := NSEC(e.Name(), names[(ln+i)%ln], s.ttl, append(e.Types(), dns.TypeNS, dns.TypeSOA, dns.TypeNSEC, dns.TypeRRSIG))
+			z.Insert(nsec)
+		} else {
+			nsec := NSEC(e.Name(), names[(ln+i)%ln], s.ttl, append(e.Types(), dns.TypeNSEC, dns.TypeRRSIG))
+			z.Insert(nsec)
+		}
+
+		for t, rrs := range zrrs {
+			if t == dns.TypeRRSIG {
+				continue
+			}
+			for _, pair := range s.keys {
+				rrsig, err := pair.signRRs(rrs, s.origin, s.ttl, s.inception, s.expiration)
+				if err != nil {
+					return err
+				}
+				e.Insert(rrsig)
+			}
+		}
+		i++
+		return nil
+	})
+	return z, err
+}
+
+// resign checks if the signed zone exists, or needs resigning.
+func (s *Signer) resign() error {
+	signedfile := filepath.Join(s.directory, s.signedfile)
+	rd, err := os.Open(signedfile)
+	if err != nil && os.IsNotExist(err) {
+		return err
+	}
+
+	now := time.Now().UTC()
+	return resign(rd, now)
+}
+
+// resign will scan rd and check the signature on the SOA record. We will resign on the basis
+// of 2 conditions:
+// * either the inception is more than 6 days ago, or
+// * we only have 1 week left on the signature
+//
+// All SOA signatures will be checked. If the SOA isn't found in the first 100
+// records, we will resign the zone.
+func resign(rd io.Reader, now time.Time) (why error) {
+	zp := dns.NewZoneParser(rd, ".", "resign")
+	zp.SetIncludeAllowed(true)
+	i := 0
+
+	for rr, ok := zp.Next(); ok; rr, ok = zp.Next() {
+		if err := zp.Err(); err != nil {
+			return err
+		}
+
+		switch x := rr.(type) {
+		case *dns.RRSIG:
+			if x.TypeCovered != dns.TypeSOA {
+				continue
+			}
+			incep, _ := time.Parse("20060102150405", dns.TimeToString(x.Inception))
+			// If too long ago, resign.
+			if now.Sub(incep) >= 0 && now.Sub(incep) > DurationResignDays {
+				return fmt.Errorf("inception %q was more than: %s ago from %s: %s", incep.Format(timeFmt), DurationResignDays, now.Format(timeFmt), now.Sub(incep))
+			}
+			// Inception hasn't even start yet.
+			if now.Sub(incep) < 0 {
+				return fmt.Errorf("inception %q date is in the future: %s", incep.Format(timeFmt), now.Sub(incep))
+			}
+
+			expire, _ := time.Parse("20060102150405", dns.TimeToString(x.Expiration))
+			if expire.Sub(now) < DurationExpireDays {
+				return fmt.Errorf("expiration %q is less than: %s away from %s: %s", expire.Format(timeFmt), DurationExpireDays, now.Format(timeFmt), expire.Sub(now))
+			}
+		}
+		i++
+		if i > 100 {
```

Comment:
Why specifically 100 records?

Reply:
it should be 1, but the RFC doesn't mandate the SOA being first. 100 seemed like a crazy enough number that makes sense? Open to suggestions.
(I'll add comment on why the current value)

Reply:
100 seems reasonable high to catch the edge cases of the SOA not being first I agree.
Comment for the reasoning sounds good. 100 were too arbitrary without a comment in my view.